add basic gonum/plot compatibility

kevinschoon opened the pull request:
Hi -- 

I've written a small compatibility layer between the [gonum/plot](https://github.com/gonum/plot) package and QFrame. It allows for the construction of types that satisfy several interfaces defined in `gonum.org/v1/plot/plotter` backed by QFrame views. The package might also be expanded by creating a higher level interface that accepts QFrame types directly and possibly return errors.

Note that I have only tested this code with a small subset of plot types defined in Gonum at this point. 

Any feedback is appreciated!

sbinet commented on contrib/gonum/plotter.go:
```diff
@@ -0,0 +1,155 @@
+package gonum
```
not completely sure calling this package `gonum` is the best move.
it's more about plotting, right?

sbinet replied:
qplot, or qfplot ?

kevinschoon replied:
Brilliant! I went with `qfplot`.

sbinet commented on examples/plotting.go:
```diff
@@ -0,0 +1,45 @@
+	})
+}
+
+func main() {
```
perhaps turn this into a real `Example` instead of a binary?
this way we run it as part of `qframe`'s tests.

sbinet commented on examples/plotting.go:
```diff
@@ -0,0 +1,45 @@
+}
+
+func main() {
+	plt, _ := plot.New()
```
as it's an example, we should add real (yet minimal) error handling.
```go
p, err := plot.New()
if err != nil {
    log.Fatal(err)
}
```

sbinet commented on examples/plotting.go:
```diff
@@ -0,0 +1,45 @@
+	canvas, _ := draw.NewFormattedCanvas(15*vg.Inch, 5*vg.Inch, "svg")
+	plt.Draw(draw.New(canvas))
+	// Write the SVG to stdout
+	canvas.WriteTo(os.Stdout)
```
please check the returned error.

also, perhaps, instead of printing the SVG to stdout, write it out to a file and test it against some reference file?
we could add the reference file to the repo and start a nice gallery as well :)

tobgu replied:
Perhaps these function calls could be replaced with the `Must...`-equivalents if those are merged before this PR?

I really like the idea of a small reference lib of SVGs with corresponding test cases that are run as part of the CI process!

sbinet commented on contrib/gonum/plotter.go:
```diff
@@ -0,0 +1,155 @@
+// defined in gonum.org/v1/plot/plotter. It accepts
+// any of the predefined LabelFunc methods in this
+// package or a custom function many be specified.
+type Labeller struct {
```
perhaps add a `plotter_test.go` file where we check we actually implement the correct interface?

```go
package gonum_test

var (
    _ plotter.Labeller = (*gonum.Labeller)(nil)
)
```

kevinschoon replied:
Added this in my most recent commit

sbinet commented on examples/plotting.go:
```diff
@@ -0,0 +1,45 @@
+	plt, _ := plot.New()
+	qf := Sin(100, 2*math.Pi)
+	// Get the views for X and Y columns
+	xview, _ := qf.FloatView("X")
```
error handling? (remember, it's an example that might be copy-pasted multiple times...)

tobgu commented:
Thanks a ton Kevin! Thanks @sbinet as well for helping out reviewing this!

I'll have a closer look in the coming days.

tobgu commented on contrib/gonum/qfplot/plotter.go:
```diff
@@ -0,0 +1,155 @@
+// Labeller implements the Labeller interface
+// defined in gonum.org/v1/plot/plotter. It accepts
+// any of the predefined LabelFunc methods in this
+// package or a custom function many be specified.
```
Picky: many -> may?

kevinschoon commented:
FYI I have started making several new additions to the API to build a better integration with the `gonum/plot` library. This change set will probably take me a bit longer than expected but any feedback would be helpful.

You can checkout an example [jupyter notebook](https://github.com/kevinschoon/qframe/blob/gonum/contrib/gonum/qfplot/examples/QFrame%20Gonum%20Compatibility.ipynb) here for a sample. Unfortunately Github doesn't seem to render the notebook too nicely!

tobgu commented:
Very cool Kevin! I'll review the code later for any feedback I might have.

kevinschoon commented:
Hi @tobgu I made several changes to the API and added a new example [here](https://github.com/kevinschoon/qframe/blob/gonum/contrib/gonum/qplot/examples/temperatures.go). I decided to remove the Jupyter notebook example for the moment and we can do that in another PR.

Thanks!

kevinschoon commented:
@sbinet @tobgu I addressed all of the items you mentioned and added a few missing interfaces between gonum and my implementation. Feel free to leave more feedback.

Thanks!

tobgu commented on contrib/gonum/qplot/plotter.go:
```diff
@@ -241,3 +241,81 @@ func MustNewXYZer(x, y, z string, qf qframe.QFrame) XYZer {
+	return XErrorer{low: lowFn, high: highFn}, nil
+}
+
+// NewXErrorer returns a new XErrorer for the values in
```
NewXErrorer -> MustNewXErrorer

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+// configures a BoxPlot after creation.
+type BoxPlotConfig func(*plot.Plot, *plotter.BoxPlot)
+
+// BotPlot returns a new PlotterFunc that plots a BoxPlot.
```
BotPlot -> BoxPlot

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+	return func(plt *plot.Plot) (plot.Plotter, error) {
+		pltr, err := plotter.NewBoxPlot(w, loc, values)
+		if err != nil {
+			return nil, err
```
Should this error be propagated rather than just returned?

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+	return func(plt *plot.Plot) (plot.Plotter, error) {
+		pltr, err := plotter.NewLabels(labeller)
+		if err != nil {
+			return nil, err
```
Same question as above with regards to error propagation.

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+// configures a QuartPlot after creation.
+type QuartConfig func(*plot.Plot, *plotter.QuartPlot)
+
+// QuartPlot returns a new PloterFunc that plots a QuartPlot.
```
PloterFunc -> PlotterFunc

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+	return func(plt *plot.Plot) (plot.Plotter, error) {
+		pltr, err := plotter.NewQuartPlot(loc, values)
+		if err != nil {
+			return nil, err
```
Propagation?

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+	return func(plt *plot.Plot) (plot.Plotter, error) {
+		pltr, err := plotter.NewYErrorBars(errorBars{XYer: xyer, YErrorer: yerr})
+		if err != nil {
+			return nil, err
```
Propagation?

tobgu commented on contrib/gonum/qplot/plotter_funcs.go:
```diff
@@ -99,12 +99,107 @@ func ScatterPlotter(xyer plotter.XYer, cfg ScatterConfig) PlotterFunc {
+	return func(plt *plot.Plot) (plot.Plotter, error) {
+		pltr, err := plotter.NewXErrorBars(errorBars{XYer: xyer, XErrorer: xerr})
+		if err != nil {
+			return nil, err
```
Propagation?

kevinschoon commented:
Updated my PR based on your feedback. I feel all the basic functionality is here to be useful but we could still improve with some more documentation / examples and possibly some corner cases with different chart types I haven't considered.

Thanks!
